teos/logging_server.py

```diff
@@ -0,0 +1,150 @@
+
+    where values that are not present are omitted. See unit tests for a more precise specification.
+    """
+
+    sio = StringIO()
+
+    ts = event_dict.pop("timestamp", None)
+    if ts:
+        sio.write(str(ts) + " ")
+
+    component = event_dict.pop("component", None)
+    if component:
+        sio.write("[" + component + "] ")
+
+    event = _repr(event_dict.pop("event"))
+
+    sio.write(event)
+
+    # Represent all the key=value elements still in event_dict
+    key_value_part = ", ".join(key + "=" + _repr(event_dict[key]) for key in sorted(event_dict.keys()))
+    if len(key_value_part) > 0:
+        sio.write("  (" + key_value_part + ")")
+
+    return sio.getvalue()
+
+
+class LogRecordStreamHandler(socketserver.StreamRequestHandler):
+    """
+    Handler for a streaming logging request. Sends to the logger any received log message, after some preprocessing.
+    """
+
+    # Taken almost Verbatim from Python's logging cookbook.
+    def handle(self):
+        """
+        Handle multiple requests - each expected to be a 4-byte length,
+        followed by the LogRecord in pickle format. Logs the record
+        according to whatever policy is configured locally.
+        """
+
+        while True:
+            chunk = self.connection.recv(4)
+            if len(chunk) < 4:
+                break
+            slen = struct.unpack(">L", chunk)[0]
+            chunk = self.connection.recv(slen)
+            while len(chunk) < slen:
+                chunk = chunk + self.connection.recv(slen - len(chunk))
+            obj = pickle.loads(chunk)
+            record = logging.makeLogRecord(obj)
+            self.handle_log_record(record)
+
+    def unPickle(self, data):
+        return pickle.loads(data)
+
+    def handle_log_record(self, record):
+        """
+        Processes log records received via the socket. The record's ``msg`` field is expected to be an encoded ``dict``
+        produced by structlog. The ``dict`` is encoded to a string using ``encode_event_dict`` and sent to the logger
+        with the name specified in the record.
+
+        Args:
+            record (:obj:`logging.LogRecord`): a log record.
+        """
+
+        event_dict = json.loads(record.msg)
+        message = encode_event_dict(event_dict)
+
+        logger = logging.getLogger(record.name)
+        logger.log(record.levelno, message, exc_info=record.exc_info, stack_info=record.stack_info)
+
+
+class LogRecordSocketReceiver(socketserver.ThreadingTCPServer):
+    """
+    Simple TCP socket-based logging receiver.
+    """
+
+    allow_reuse_address = True
+
+    def __init__(self, host="localhost", port=TCP_LOGGING_PORT, handler=LogRecordStreamHandler):
+        socketserver.ThreadingTCPServer.__init__(self, (host, port), handler)
+        self.abort = 0
+        self.timeout = 1
+        self.logname = None
+
+    def serve_until_stopped(self):
+        abort = 0
+        while not abort:
```

Comment:
My understanding is that this is stopped when self.abort is set to `1` after a new log is read, given that if that does not happen abort is always `0`.

I think this could be better handled with a signal in the same way we're handling other processes.

Reply:
I tried with signals, but didn't quite work; instead, to keep things simple (the logging process is the last service that should stop anyway), I changed `serve_until_stopped` to `serve_forever` in 482b429, and removed the `abort` flag from `LogRecordSocketReceiver`.

Reply:
Since you're overwriting the base class method, stick to the method signature:

```
def serve_forever(self, poll_interval=0.5)
```

Reply:
Oops, I didn't realize it was overwriting a method. Fixed.